synapse/lib/stormlib/model.py

```diff
@@ -692,3 +693,203 @@ async def _lock(self, name, locked):
+                    verbs.add(verb)
+
+                n1proto = await editor.getNodeByBuid(s_common.uhex(n1iden))
+                if n1proto is not None:
+                    await n1proto.addEdge(verb, dstiden)
+
+    async def _methCopyTags(self, src, dst, overwrite=False):
+
+        if not isinstance(src, s_node.Node):
+            raise s_exc.BadArg(mesg='$lib.model.migration.copyTags() source argument must be a node.')
+        if not isinstance(dst, s_node.Node):
+            raise s_exc.BadArg(mesg='$lib.model.migration.copyTags() dest argument must be a node.')
+
+        overwrite = await s_stormtypes.tobool(overwrite)
+
+        snap = self.runt.snap
+
+        async with snap.getEditor() as editor:
+
+            proto = editor.loadNode(dst)
+
+            for name, valu in src.tags.items():
+                self.runt.layerConfirm(('node', 'tag', 'add', *name.split('.')))
+                await proto.addTag(name, valu=valu)
+
+            for tagname, tagprops in src.tagprops.items():
+                for propname, valu in tagprops.items():
+                    if overwrite or not dst.hasTagProp(tagname, propname):
+                        await proto.setTagProp(tagname, propname, valu) # use tag perms
+
+@s_stormtypes.registry.registerLib
+class LibModelMigrations(s_stormtypes.Lib):
+    '''
+    A Storm library for selectively migrating nodes.
+    '''
+    _storm_locals = (
+        {'name': 'riskHasVulnToVulnerable', 'desc': '''
+            Create a risk:vulnerable node from the provided risk:hasvuln node.
+
+            Edits will be made to the risk:vulnerable node in the current write layer.
+
+            If multiple vulnerable properties are set on the risk:hasvuln node
+            multiple risk:vulnerable nodes will be created (each with a unique guid).
+            Otherwise, a single risk:vulnerable will be created with the same guid
+            as the provided risk:hasvuln node.
+
+            Tags, tag properties, edges, and node data will be copied
+            to the risk:vulnerable node. However, existing tag properties and
+            node data will not be overwritten.
+        ''',
+        'type': {'type': 'function', '_funcname': '_storm_query',
+                 'args': (
+                      {'name': 'n', 'type': 'node', 'desc': 'The risk:hasvuln node to migrate.'},
+                 ),
+                 'returns': {'type': 'list', 'desc': 'A list of risk:vulnerable nodes.'}}},
+    )
+    _storm_lib_path = ('model', 'migration', 's')
+
+    _storm_query = '''
+        function riskHasVulnToVulnerable(n) {
+
+            if (not $n.isform(risk:hasvuln)) {
+                $lib.raise(BadArg, `riskHasVulnToVulnerable() only accepts risk:hasvuln nodes, not {$n.form()}.`)
+            }
+
+            $ret = ([])
+            $links = ({})
+
+            $vuln = $n.props.vuln
+            if (not $vuln) { return() }
+
+            for $prop in (hardware, host, item, org, person, place, software, spec) {
```

Comment:
If someone has created extended properties on `risk:hasvuln` we would be missing them. We could inspect the model to identify any extended forms and migrate them as well?

Reply:
I started there, but then decided against it since there is no guarantee those extended props should actually be considered risk:vulnerable:node.  I can add a note that extended props will not be migrated.

Reply:
Adding a note to the description would be sufficient.

Reply:
Updated note